Feature/autogenerate attack tactics visualization 95

tmendonca28 opened the pull request:
### What does this PR do?

Adds the functionality to generate stylized HTML coverage matrices for each platform with Columns: ATT&CK Tactics and Rows: Stratus Red Team techniques.
Integrates the GenerateCoverageMatrices function into the generate-docs.go workflow so that platform-specific matrices are generated by running `make docs`

- Tested the output for correctness and styling (e.g., AWS.html, GCP.html).
- Added hover effects and color coding to improve readability and likeness to Stratus Red Team color scheme.
- Ensured no breaking changes to the existing codebase.
- Verified that the generated HTML includes the favicon for branding.

### Motivation

Resolves #95
This PR addresses the need for automatically generating stylized coverage matrices for all platforms currently covered by Stratus Red team, improving documentation usability and visibility for maintainers and users.

### Checklist
N/A

### Sample Screenshot of HTML Table
<img width="1501" alt="Screenshot 2025-01-01 at 23 58 35" src="https://github.com/user-attachments/assets/00145e55-0a25-4ce1-999e-2940edfb4a57" />

christophetd commented:
Looks really good, thanks for giving it a stab! Do you think this is worth integrating it into the main docs website? Perhaps under a new "MITRE ATT&CK Coverage" page

tmendonca28 commented:
Thank you for the feedback! I think integrating the coverage matrix into the main docs website under a new “MITRE ATT&CK Coverage” page would be an excellent idea. It would potentially provide greater visibility and usability for users.

I’d be happy to explore this further and draft an implementation plan. Would you prefer a static integration, or should we explore automating updates to the page as new techniques are added? I personally would like the latter but I'm not quite sure (yet) how to do it but will explore the codebase further. 

Looking forward to your thoughts!

christophetd commented:
I think an automatic option that nukes and regenerates a static file (e.g. in `docs/attack-techniques/mitre-attack.md`/html) would be perfect, perhaps split by platform.

Thoughts?

tmendonca28 commented:
Thanks @christophetd  for the above clarification. I think regenerating a static file automatically seems feasible. Based on your input above, here is a draft implementation plan I propose:

- **File location/structure**: The static file will contain MITRE ATT&CK coverage tables split by platform, probably with a main title (MITRE ATT&CK Coverage by Platform). The static file will be located at **docs/attack-techniques/mitre-attack.md/html**. 
- **'Nuking' process**: Delete previous version of the file and regenerate it from scratch whenever needed. I would like to do this in the existing **generate-docs** workflow so it can be regenerated when running `make docs`.
- **File output**: I think I will work on doing my best to create a markdown file for this for integration into the main docs website but might prove a bit finicky.

Let me know if you'd like any adjustments. I'm happy to start working on this!

christophetd commented:
That sounds absolutely great!

christophetd commented:
@tmendonca28 Just checking if you're still on it or if you need me to take it from there

tmendonca28 commented:
Hi @christophetd . Thanks a lot for your patience, I'd still like to finish up on this, will push something out by EOW latest.

tmendonca28 commented:
@christophetd  Thanks for the suggestion to simply throw the HTML into the markdown file, it worked pretty well. Here are a summary of the changes I've made for your review:

- Updated `generate-docs.go`
- Updated `generate-coverage-matrices.go`
  - Modified the function to generate mitre-attack.md instead of the individual HTML files, embedding the full HTML inside the markdown file.
  - The output file is placed in `docs/attack-techniques/mitre-attack.md`
- Sorted MITRE ATT&CK Tactics in the coverage matrices in the correct order of execution
  - The columns in the resultant tables are **now arranged according to the MITRE ATT&CK framework order** for techniques that are present i.e. Recon  to Impact as can be found [here](https://attack.mitre.org/). 
  - I noticed the order was a bit mixed when pulling it from the map and thought it would be useful and easier for users to quickly assess overall ATT&CK techniques covered by SRT in a logical attack sequence. You can see screenshots of the outcome below:
  ![image](https://github.com/user-attachments/assets/4139c144-5d56-4466-983a-37b00e4d7c84)
  ![image](https://github.com/user-attachments/assets/e1338768-410b-473a-b3c9-ccbbe4c5c276)

- Updated `mkdocs.yml'
  - Added MITRE ATT&CK Coverage Matrices to the file so it can be found in the main docs website.
  - Each time `make docs` is run now deletes and regenerates `mitre-attack.md` ensuring the most up-to-date coverage is available. (aka the 'Nuking' process you mentioned above)

Would love your thoughts on this approach! Let me know if you prefer any refinements.

christophetd commented:
This looks amazing, I love it! Small HTML issue on http://127.0.0.1:8000/attack-techniques/mitre-attack-coverage-matrices/ I think (font family is messed up) but that's a minor thing to fix

The main UX painpoint is that horizontal scrolling make the matrix hard to see (even on a 27" screen): 

<img width="1639" alt="image" src="https://github.com/user-attachments/assets/912f4002-21e5-4074-8713-7ae254db2413" />


Any suggestions on how to solve this?

Again, super cool job that I'm eager to merge

tmendonca28 commented:
Hi @christophetd, thanks a lot for the feedback! 
I've managed to sort out the font-family issue by removing the body tag I had placed initially in the HTML table rendering.

WRT this comment _"The main UX painpoint is that horizontal scrolling make the matrix hard to see (even on a 27" screen)"_ there are 2 options I can think of:

- Creating a sort of scrollable table i.e. update the table styling so that it scrolls within a container and hence prevents the full-page scrolling. This will not ideally fix the horizontal scrolling issue but tone it down a bit.
- Creating toggle-able full-screen mode to improve visibility maybe via a "View Full Screen" button to expand tables while keeping the default compact layout. This might be good as it'll expand the table and probably allow a user to see the entire length of the table.

Please let me know what you think of these and I can try my hand at enhancing usability.

WRT this "we should probably make the attack techniques in the table clickable and hyperlink to the appropriate technique":

- I think this is a great idea and I would love to implement this as it'll really make the table more useful. I can definitely work on this assuming the technique documentation follows a consistent URL structure (which I think it does).  Give me till the weekend when I get some time and I will push a commit to aid with this.

christophetd commented:
@tmendonca28 Thanks! I added some styling that should look slightly better as well as hyperlinks. If that works for you, I'm happy to merge as is!

tmendonca28 commented:
@christophetd  Thanks a lot for your patience and adding in the styling and links. I did not realize you could manipulate it that easily using the Sprintf!

I made 2 changes that I think would improve consistency/UX 😄 :

- Firstly, I capitalized the Platform names since I they would look better and more consistent than having a mix of "all capital letters" for AWS then "small letters" for kubernetes for example (might be my OCD too 😆 ). Here is a screenshot:
![image](https://github.com/user-attachments/assets/4830c5ca-5a83-4266-ba2e-2f0d367d0bbe)

- Secondly, I couldn't scroll horizontally with my mouse and went ahead and added scrollable div around our "platform tables". I think this really enhances the UX and usability for the user when they are scrolling across the table. Please LMKWYT. 
Here are screenshots of this change:
![image](https://github.com/user-attachments/assets/d345187d-98d5-41ff-809b-4236205758b2)
![image](https://github.com/user-attachments/assets/cdf407c6-8467-4e48-8b2e-1970a830de0f)

Other than that, I am super happy for you to merge this, and I really appreciate your patience once again 😃
